### ISSUE TITLE
Preflight OPTIONS requests fail with 408 - Request timeout

### ISSUE THREAD

**Issue by vejja** [supabase/edge-runtime#50]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

When using `supabase functions serve` to serve all functions simultaneously, the pre-flight 'OPTIONS' request initiated by a browser agent fails.
The server returns Status Code `408 - Request timeout` with a message body of `{"msg":"Request could not be processed by the server because it timed out or an error was thrown."}`

**The bug is triggered by the fact that the OPTIONS body is empty**

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Use code snippet at [https://supabase.com/docs/guides/functions/cors#recommended-setup](https://supabase.com/docs/guides/functions/cors#recommended-setup) to setup CORS headers in the `serve` function
2. Send `OPTIONS` request to edge function

## Expected behavior

The server should return `200` with the CORS headers set appropriately.

## Screenshots

![Capture d’écran 2023-04-13 à 15 22 38](https://user-images.githubusercontent.com/7295259/231773795-0576af7b-09b5-480f-a723-61bddf9889f0.png)
![Capture d’écran 2023-04-13 à 15 23 28](https://user-images.githubusercontent.com/7295259/231773842-7c314f92-94db-4155-8416-2822d0883cf3.png)

## System information

- OS: macOS
- Browser (if applies) : chrome. Also tested outside browser (Postman)
- Version of supabase-js: 2.20.0
- Version of Node.js: not relevant

## Additional context

1. If serving the function individually via `supabase functions serve <function-name>`, the bug does not happen. It only happens when serving all functions simultaneously via `supabase functions serve`

2. If calling the function with the `POST` method directly, the bug does not happen. 

3. Inserting `console.log()` statements everywhere in the edge function code shows that the script reaches the point where it has to send the response to the OPTIONS request, then crashes silently.

4. If you put some dummy data into the body of the `OPTIONS` request, the bug does not happen:
![Capture d’écran 2023-04-13 à 15 24 00](https://user-images.githubusercontent.com/7295259/231774452-28ca7111-9454-45ca-ad1a-722c4a8eaabc.png)

*Unfortunately a browser-generated pre-flight request always has an empty body*

To me it looks like when I call `return new Response('ok', { headers: corsHeaders })` inside the edge function, this triggers the bug because probably the server-side is still waiting for the client to send its body and it does not want to reply with the Response before it can close the incoming connection - but I may be mistaken on my interpretation.

**In conclusion, it looks like there may be some code inside edge-runtime that tries to enforce a non-empty body on an `OPTIONS` request ?**

**Comment by vejja** [comment]
Additional info : 
- broken somewhere between supabase-cli 1.47.5 and 1.50.4
- not solved as of supabase-cli 1.50.8

Linking to supabase-cli as it might be a better place to report

**Comment by nminchow** [comment]
I'm getting this as well. I've also noticed the following behavior. If I send:
```
OPTIONS http://localhost:54321/functions/v1/function-name
```

with no headers, the response fails. If I include a body in the options request:

```
OPTIONS http://localhost:54321/functions/v1/function-name

{
  "name": "test"
}
```

The response succeeds.

**Comment by joshmarnold** [comment]
I'm also seeing this. Posted about it on [Discord](https://discord.com/channels/839993398554656828/1097177678412533903/1097177678412533903)

**Comment by vejja** [comment]
@soedirgo : Not sure whether this is relevant to edge-runtime, but I think DELETE requests also have an empty body.
Anyways many thanks for the fix 🚀

**Comment by Aaron3219** [comment]
> Additional info :
> 
> * broken somewhere between supabase-cli 1.47.5 and 1.50.4
> * not solved as of supabase-cli 1.50.8
> 
> Linking to supabase-cli as it might be a better place to report

supabase-cli 1.48.1 is still working. So it must be between 1.48.1 and 1.50.4

**Comment by andreespirela** [comment]
<img width="1363" alt="image" src="https://user-images.githubusercontent.com/60560109/234606202-d7e13d94-b0ca-4dd5-8f27-b96bf694d661.png">

cc @laktek I'm not able to reproduce this bug in the latest `main`. Can we close?

**Comment by laktek** [comment]
@andreespirela Can you try other HTTP verbs like DELETE and confirm everything works without a request body?

**Comment by andreespirela** [comment]
@laktek Getting this instead

```
{
    "msg": "Error: channel closed"
}
```

Investigating.

**Comment by andreespirela** [comment]
We are weirdly handling when a request has a body
```
    async fetch(req) {
        const { method, url, headers, body, bodyUsed } = req;

        const headersArray = Array.from(headers.entries());
        const hasBody = false;

        const userWorkerReq = {
            method,
            url,
            headers: headersArray,
            hasBody,
        };

        const { requestRid, requestBodyRid } = await ops.op_user_worker_fetch_build(userWorkerReq);

        // stream the request body
        if (hasBody) {
            let writableStream = writableStreamForRid(requestBodyRid);
            body.pipeTo(writableStream);
        }

        const res = await core.opAsync("op_user_worker_fetch_send", this.key, requestRid);
        const bodyStream = readableStreamForRid(res.bodyRid);
        return new Response(bodyStream, {
            headers: res.headers,
            status: res.status,
            statusText: res.statusText
        });
    }
```
`hasBody` 
cc @laktek

**Comment by joshmarnold** [comment]
I'm confused how this could be fixed as I thought that #56 was supposed to fix this and it hasn't been merged. In either case, could we pls get an update when its released?

**Comment by vejja** [comment]
> We are weirdly handling when a request has a body

Agree - See https://github.com/supabase/edge-runtime/pull/69#issuecomment-1525346087

**Comment by DeanWard** [comment]
I feel like i'm missing something obvious here but I cannot work out how to get my local supabase to use the latest version. 

I've updated supabase CLI via homebrew but it's still pulling down edge runtime v1.2.12 from docker, i've manually deleted all the images and allowed supabase to re pull them but it's still pulling the old version.

**Comment by laktek** [comment]
@DeanWard We haven't updated Supabase CLI to use the latest Edge Runtime yet. Expect a new release today.

**Comment by laktek** [comment]
Supabase CLI 1.52.3+ should include the Edge Runtime with the fix. cc: @DeanWard @vejja

**Comment by CareTiger** [comment]
@laktek Lakshan - i updated my cli this morning and i only have 1.52.2. when will 1.52.3 be release? thank

**Comment by vejja** [comment]
Wonderful - fixed !
Thanks guys

**Comment by laktek** [comment]
@CareTiger 1.52.3 should already be available. Can you try upgrading again?

**Comment by CareTiger** [comment]
still doesnt work. ```serving the request with /home/deno/functions/hello-world
Error: channel closed
    at async Function.create (ext:sb_user_workers/user_workers.js:75:21)
    at async Server.<anonymous> (file:///home/deno/main/index.ts:94:24)
    at async Server.#respond (https://deno.land/std@0.182.0/http/server.ts:220:24)```

**Comment by CareTiger** [comment]
@laktek  supabase --version
1.52.3

**Comment by laktek** [comment]
@CareTiger Can you share the request you are making? Also, the code for the `hello-world` function?

**Comment by CareTiger** [comment]
I have filed a ticket. my project reference is zkcmpwzbigjgwiesauro. here is the code ```
// @ts-ignore: module not found
import { serve } from "server";
// @ts-ignore: module not found
import { corsHeaders } from "cors";

// console.log("Hello from Functions!");

serve(async (req: Request) => {
    if (req.method === "OPTIONS") {
        return new Response("ok", { headers: corsHeaders });
    }
    const { pathname: path, searchParams: query } = new URL(req.url);
    console.log(path);
    console.log(req.url);
    // get the payload from the request
    const payload = await req.json();

    console.log(payload.user_id);
    
    return new Response(JSON.stringify(`Hello ${payload.user_id}!`), {
        headers: {
            ...corsHeaders,
        },
    });
});```

**Comment by CareTiger** [comment]
@laktek others are facing the same issue - https://discord.com/channels/839993398554656828/1097177678412533903/1103052379877818419

**Comment by laktek** [comment]
An update: Looks like @CareTiger's issue was caused due to a bug in CLI handling import maps. Will do a fix for it soon (check the above Discord thread for more details)
